### PR TITLE
Misc patches leading to the term model

### PR DESCRIPTION
**Pull request by robdockins**
A variety of minor changes and fixes required by the term model, or discovered while working on it.

### CODE REVIEW

**Comment by brianhuffman** [comment]
The new term defaulting functionality might be useful for #317.

File: saw-core/src/Verifier/SAW/Simulator.hs [hunk 1]
```diff
@@ -142,10 +142,10 @@ evalTermF cfg lam recEval tf env =
                                     if inBitSet 0 (looseVars t2) then
                                       pure (VDependentPi (\x -> toTValue <$> lam t2 ((x,v) : env)))
                                     else
-                                      do val <- delay (panic "evalTermF"
-                                                         ["nondependent Pi type forced its value"
-                                                         , showTerm (Unshared tf)])
-                                         VNondependentPi . toTValue <$> lam t2 ((val,v) : env)
+                                      do -- put dummy values in the environment; the term should never reference them
+                                         let val = ready VUnit
+                                         let tp  = VUnitType
```
Comment (brianhuffman) [inline saw-core/src/Verifier/SAW/Simulator.hs:147]: Putting a term with the wrong type in the environment seems a bit icky to me. (Granted, hiding a `panic` in there was already icky.) Can you think of any cleaner fixes for this? Perhaps the environment could be a list of `Maybe`s, and we could just insert a `Nothing`?

File: saw-core/src/Verifier/SAW/SharedTerm.hs [hunk 2]
```diff
@@ -2359,6 +2360,28 @@ scAbstractExts sc exts x = loop (zip (inits exts) exts)
...
 
+-- | Abstract over the given list of external constants by wrapping
+--   the given term with lambdas and replacing the external constant
+--   occurrences with the appropriate local variables.  However,
+--   the term will be eta-collapsed as far as possible, so unnecessary
+--   lambdas will simply be omitted.
+scAbstractExtsEtaCollapse :: SharedContext -> [ExtCns Term] -> Term -> IO Term
+scAbstractExtsEtaCollapse sc = \exts tm -> loop (reverse exts) tm
+  where
+    -- we eta-collapesed all the variables, nothing more to do
```
Comment (brianhuffman) [inline saw-core/src/Verifier/SAW/SharedTerm.hs:2406]: "collapesed"

File: saw-core/src/Verifier/SAW/Simulator.hs (hunk 1 above)
Comment (robdockins) [reply saw-core/src/Verifier/SAW/Simulator.hs:147]: I guess we could do that, but it adds an additional indirection and pattern match on the common case.

I claim we can put any type into the context that we like, provided the variable doesn't appear in the term; and that this follows as a pretty standard consequence of weakening.

Comment (brianhuffman) [reply saw-core/src/Verifier/SAW/Simulator.hs:147]: Yeah, I guess the need to do a pattern match is not nice. I suppose it's fine how it is.

**Review by brianhuffman — APPROVED** [review]

**Pull request merged**